src/python/pants/backend/python/rules/pex.py

```diff
@@ -15,18 +15,65 @@
   DirectoriesToMerge,
   DirectoryWithPrefixToAdd,
 )
-from pants.engine.isolated_process import ExecuteProcessResult, MultiPlatformExecuteProcessRequest
+from pants.engine.isolated_process import (
+  ExecuteProcessRequest,
+  ExecuteProcessResult,
+  MultiPlatformExecuteProcessRequest,
+)
+from pants.engine.legacy.structs import PythonTargetAdaptor, TargetAdaptor
 from pants.engine.platform import Platform, PlatformConstraint
-from pants.engine.rules import optionable_rule, rule
+from pants.engine.rules import RootRule, optionable_rule, rule
 from pants.engine.selectors import Get
+from pants.util.strutil import create_path_env_var
+
+
+@dataclass(frozen=True)
+class PexRequirements:
+  requirements: Tuple[str] = ()
+
+  @classmethod
+  def create_from_adaptors(cls, adaptors: Tuple[TargetAdaptor], additional_requirements: Tuple[str]) -> 'PexRequirements':
+    all_target_requirements = []
+    for maybe_python_req_lib in adaptors:
+      # This is a python_requirement()-like target.
+      if hasattr(maybe_python_req_lib, 'requirement'):
```

Comment:
Not for this PR, but in general RE #4535 it would be nice if this were maybe detected by superclass (and then a property access) rather than by grabbing around for raw field names...

Reply:
Absolutely, but yeah I don't want to handle that chunk of work until this PR is merged.